sdk/go/common/resource/plugin/provider-implementers-guide.md

```diff
@@ -0,0 +1,609 @@
+             detected no changes, or does not support detailed diff detection. Providers
+             should return `Some` for this value if there are any entries in
+             `detailedDiff`; otherwise they should return `None` to indicate no
+             difference. If a provider returns `Unknown` for this value, it is the
+             responsibility of the client to determine whether or not differences exist
+             by comparing the resource's last recorded _inputs_ with its current inputs.
+
+In addition, the following properties should be returned for compatibility with older
+clients:
+
+- `replaceKeys`: the list of top-level input property names with changes that require that the
+                 resource be replaced.
+- `stableKeys`: the list of top-level input property names that did not change and
+                top-level output properties that are guaranteed not to change.
+- `changedKeys`: the list of top-level input property names that changed.
+
+If a provider is unble to compute a diff because its configuration contained
+[`Unknown`s](#unknowns), it can return an error that indicates as such. The client should
+conservatively assume that the resource must be updated and warn the user.
+
+#### Detailed Diffs
+
+A detailed diff is a map from [property paths](#property-paths) to change kinds that
+describes the differences between the actual and desired state of a resource and the
+operations necessary to reconcile the two.
+
+Each entry in a detailed diff has a change kind that describes how the value of and
+input property differs, whether or not the difference requires replacement, and which old
+value was used for determining the difference. The core change kinds are:
+
+- `Add`, which denotes an `Object` property or `Array` element that was added
+- `Update`, which denotes an `Object` property or `Array` element that was updated
+- `Delete`, which denotes an `Object` property or `Array` element that was removed
+
+Each of these core kinds is paramaterized on whether or not the change requires
+replacement and whether the old value of the property should was read from the
+resource's old input `Object` or old state `Object`.
+
+*TODO*: the input/output flag is a bit clumsy, as it is the only part of the system
+that implies some correspondence between input and output `Object` schemas. It was
+chosen over an approach that used old/new values due in order to remove the possibility
+of a provider accidentally revealing a secret value as part of a diff. We should
+reconsider this approach if we can find an easy way to maintain secretness.
+
+#### Create
+
+The `Create` method is responsible for creating a new instance of a resource from an
+input `Object` and returning the resource's state `Object`. `Create` may be called during
+a [preview](#preview) in order to compute a hypothetical state `Object` without actually
+creating the resource, in which case the `preview` argument will be `true`.
```

Comment:
Huh - I did know about this.  Do any of our existing providers support this?  What happens if a provider doesn't support this (as I assume it's nontrivial to support?)